src/dictionary.c

```diff
@@ -0,0 +1,103 @@
+
+/*
+ * A dictionary data structure based on tries
+ *
+ * See dictionary.h for function documentation.
+ */
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <assert.h>
+#include <unistd.h>
+#include "utils.h"
+#include "dictionary.h"
+
+/* See dictionary.h */
+dict_t* dict_new() {
+
+    dict_t *d;
+    int rc;
+
+    d = malloc(sizeof(dict_t));
+
+    if (d == NULL) {
+        error("Could not allocate memory");
+        return NULL;
+    }
+
+    rc = dict_init(d);
+    if (rc != 0) {
+        error("Could not initialize dict");
+        return NULL;
+    }
+
+    return d;
+}
+
+/* See dictionary.h */
+int dict_init(dict_t *d) {
+    assert(d != NULL);
+
+    trie_t *t = trie_new();
+    if (t == NULL) {
+        return 1;
+    }
+    d->dict = t;
+
+    return 0;
+}
+
+/* See dictionary.h */
+int dict_free(dict_t *d) {
+    assert(d != NULL);
+    assert(d->dict != NULL);
+
+    trie_free(d->dict);
+    free(d);
+
+    return 0;
+}
+
+/* See dictionary.h */
+int in_dict(char *str, dict_t *d) {
```

Comment:
All `dict` methods should start with `dict_`. `dict_in` just doesn't read right, so let's make this `dict_exists` (the reasoning is that since C lacks a proper module system, we can kinda sorta mimic modules by just prefixing method names consistently).

Also, let's make it so that all `dict` methods have the `dict_t *` as the first parameter.